CesiumAsync/src/CachingAssetAccessor.cpp

```diff
@@ -309,23 +311,24 @@ bool shouldCacheRequest(
     return false;
   }
 
-  // check if cache control contains no-store or no-cache directives
-  int maxAge = 0;
+  // Check cache control header if it exists
+  bool ignoreExpiresHeader = false;
   if (cacheControl) {
-    if (cacheControl->noStore() || cacheControl->noCache()) {
+    if (cacheControl->noStore())
       return false;
-    }
 
-    maxAge = cacheControl->maxAge();
+    // If there is a Cache-Control header with the max-age or s-maxage directive
+    // in the response, the Expires header is ignored
+    ignoreExpiresHeader =
+        cacheControl->maxAgeExists() || cacheControl->sharedMaxAgeExists();
   }
 
-  // check response header contains expires if maxAge is not specified
-  const HttpHeaders& responseHeaders = pResponse->headers();
-  if (maxAge == 0) {
-    HttpHeaders::const_iterator expiresHeader = responseHeaders.find("Expires");
-    if (expiresHeader == responseHeaders.end()) {
+  // Check older Expires header
+  if (!ignoreExpiresHeader) {
+    const HttpHeaders& headers = pResponse->headers();
+    HttpHeaders::const_iterator expiresHeader = headers.find("Expires");
+    if (expiresHeader == headers.end())
```

Comment:
I think something is missing here. What if we have these headers:

```
Cache-Control: no-cache
ETag: whatever
```

And no `Expires`.

This should be cached, and it should be revalidated with `If-None-Match`. But currently this logic will return false and it will not be stored.

Or what if we have this:

```
Cache-Control: no-cache
Last-Modified: [some date]
```

Same thing: it should be stored and revalidated by `If-Modified-Since`, but currently it wouldn't be cached at all.

Unless I'm missing something (very possible), I think the logic should be:

1. If it's `no-store`, don't store (because we're not allowed to), else...
2. If it has a `max-age` or an `Expires` that puts its expiration in the future, then store it (because it might be reusable without revalidation in the time window), else...
3. If it has an `ETag` or a `Last-Modified`, then store it (because it might be reusable with revalidation even though it's already expired)

Note that a response we previously cached for rule 2 might be still be reusable with revalidation after its expiration if it also has an `ETag` or a `Last-Modified`. But we don't need to check when deciding whether to store it.

Reply:
Committed two more unit tests to cover these no-cache examples, as well as the changed logic to cache them correctly.

Reply:
I think it covers my specific examples now, but the logic is still not quite right. I still stand by the algorithm I described above, although I should have added a step 4, "else don't store". Here's a new specific example that the current algorithm would get wrong (but this is definitely not meant to be an exhaustive inventory of cases):

```
ETag: whatever
```

No cache control, no expires. But it should still be cached because it can be revalidated by ETag later.

Reply:
>  although I should have added a step 4, "else don't store"

Ah yes, that is a non-trivial change at the end of `shouldCacheRequest`, but doable. I've added your 2 new examples as unit tests, and changed the logic to match your algorithm closer.

Thanks for tolerating this back and forth. I tend to prefer test driven development with logic like this, thus the new unit tests. But having intuitive and well thought through algorithms is important too. Hopefully we get the best of both worlds.